gcloud-java-contrib/gcloud-java-nio/src/main/java/com/google/gcloud/storage/contrib/nio/CloudStorageFileSystemProvider.java

```diff
@@ -537,8 +539,32 @@ public void createDirectory(Path dir, FileAttribute<?>... attrs) {
    */
   @Override
   public DirectoryStream<Path> newDirectoryStream(Path dir, Filter<? super Path> filter) {
-    // TODO: Implement me.
-    throw new UnsupportedOperationException();
+    CloudStoragePath cloudPath = checkPath(dir);
+    checkNotNull(filter);
+    String prefix = cloudPath.toString();
+    Iterator<Blob> iterator = storage.list(cloudPath.bucket(), Storage.BlobListOption.prefix(prefix)).iterateAll();
```

Comment:
Looks like we are only interested in names. What about restricting the fields returned by passing [BlobListOption.field()](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-storage/src/main/java/com/google/cloud/storage/Storage.java#L714)?